zerver/views/upload.py

```diff
@@ -0,0 +1,74 @@
+from django.shortcuts import redirect
+from django.http import HttpResponseForbidden
+
+from zerver.decorator import has_request_variables, REQ
+from zerver.lib.response import json_success, json_error
+from zerver.lib.validator import check_bool
+from zerver.lib.upload import get_local_file_path, upload_message_image_through_api
+from zerver.models import validate_attachment_request
+from zerver.views import get_uploaded_file, get_realm_for_filename, get_signed_upload_url
+
+from sendfile import sendfile
+
+from zerver.lib.rest import rest_dispatch as _rest_dispatch
+rest_dispatch = csrf_exempt((lambda request, *args, **kwargs: _rest_dispatch(request, globals(), *args, **kwargs)))
+
+def serve_local(request, path_id):
+    local_path = get_local_file_path(path_id)
+    if local_path is None:
+        return json_error("That file does not exist.", status=404)
+    else:
+        return sendfile(request, local_path)
+
+def serve_s3(request, user_profile, realm_id, filename, redir):
+    user_profile = request.user
+    url_path = "%s/%s" % (realm_id, filename)
+
+    if realm_id == "unk":
+        realm_id = get_realm_for_filename(url_path)
+        if realm_id is None:
+            # File does not exist
+            return json_error("That file does not exist.", status=404)
+
+    # Internal users can access all uploads so we can receive attachments in cross-realm messages
+    if user_profile.realm.id == int(realm_id) or user_profile.realm.domain == 'zulip.com':
+        uri = get_signed_upload_url(url_path)
+        if redir:
+            return redirect(uri)
+        else:
+            return json_success({'uri': uri})
+    else:
+        return HttpResponseForbidden()
+
+@has_request_variables
+def api_get_uploaded_file(request, user_profile, realm_id, filename, redir=REQ(validator=check_bool, default=True)):
+    path_id = "%s/%s" % (realm_id, filename)
+    is_authorized = validate_attachment_request(user_profile, path_id)
+
+    if is_authorized == None:
+        return json_error("That file does not exist.", status=404)
+    if not is_authorized:
+        return json_error("You are not authorized to view this file.", status=403)
+
+    if settings.LOCAL_UPLOADS_DIR is not None:
+        return serve_local(request, path_id)
+    else:
+        return serve_s3(request, user_profile, realm_id, filename, redir)
+
+@has_request_variables
+def api_upload_file(request, user_profile):
+    if len(request.FILES) == 0:
+        return json_error("You must specify a file to upload")
+    if len(request.FILES) != 1:
+        return json_error("You may only upload one file at a time")
+
+    user_file = list(request.FILES.values())[0]
+    if ((settings.MAX_FILE_UPLOAD_SIZE * 1024 * 1024) < user_file._get_size()):
+        return json_error("File Upload is larger than allowed limit")
+
+    uri = upload_message_image_through_api(request, user_file, user_profile)
+    return json_success({'uri': uri})
```

Comment:
This file has a lot of duplicated code with the existing web upload code in zerver/views/__init__.py.  Can you fix that?

How I would do it is a series of commits like this:
(1) Move the existing upload code from zerver/views/__init__.py into a new zerver/views/upload.py, with no modifications other than adjusting imports in that commit (so it's reviewable!)
(2) Do any refactoring commit(s) you want to do to clean up that code or support the API use case
(3) In this commit, add small API routes for accepting uploads, which should add very little new code.